File: Helpers/src/main/java/com/dg/helpers/DbHelper.java
package com.dg.helpers;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.Map;
import java.util.TimeZone;

public class DbHelper
{
    public static SimpleDateFormat utcIsoDateFormatter()
    {
        return DateHelper.utcIsoDateFormatter();
    }

    public static String getDateForDbFromDate(Date date, boolean escape)
    {
        if (date != null)
        {
            if (escape)
            {
                return "'" + DateHelper.utcIsoDateFormatter().format(date) + "'";
            }
            else
            {
                return DateHelper.utcIsoDateFormatter().format(date);
            }
        }
        else
        {
            if (escape)
            {
                return "null";
            }
            else
            {
                return null;
            }
        }
    }

    public static Date getDateFromDbString(String date)
    {
        if (date == null)
        {
            return null;
        }
        try
        {
            return DateHelper.utcIsoDateFormatter().parse(date);
        }
        catch (ParseException ignored)
        {
            return null;
        }
    }

    public static Float queryScalarFloat(SQLiteDatabase database, String sql)
    {
        Cursor cursor = database.rawQuery(sql, null);
        if (cursor.moveToFirst())
        {
            Float val = cursor.isNull(0) ? null : cursor.getFloat(0);
            cursor.close();
            return val;
        }
        cursor.close();
        return null;
    }

    public static Integer queryScalarInt(SQLiteDatabase database, String sql)
    {
        Cursor cursor = database.rawQuery(sql, null);
        if (cursor.moveToFirst())
        {
            Integer val = cursor.isNull(0) ? null : cursor.getInt(0);
            cursor.close();
            return val;
        }
        cursor.close();
        return null;
    }

    public static Long queryScalarLong(SQLiteDatabase database, String sql)
    {
        Cursor cursor = database.rawQuery(sql, null);
        if (cursor.moveToFirst())
        {
            Long val = cursor.isNull(0) ? null : cursor.getLong(0);
            cursor.close();
            return val;
        }
        cursor.close();
        return null;
    }

    public static Double queryScalarDouble(SQLiteDatabase database, String sql)
    {
        Cursor cursor = database.rawQuery(sql, null);
        if (cursor.moveToFirst())
        {
            Double val = cursor.isNull(0) ? null : cursor.getDouble(0);
            cursor.close();
            return val;
        }
        cursor.close();
        return null;
    }

    public static Short queryScalarShort(SQLiteDatabase database, String sql)
    {
        Cursor cursor = database.rawQuery(sql, null);
        if (cursor.moveToFirst())
        {
            Short val = cursor.isNull(0) ? null : cursor.getShort(0);
            cursor.close();
            return val;
        }
        cursor.close();
        return null;
    }

    public static Boolean queryScalarBoolean(SQLiteDatabase database, String sql)
    {
        Cursor cursor = database.rawQuery(sql, null);
        if (cursor.moveToFirst())
        {
            Boolean val = cursor.isNull(0) ? null : cursor.getInt(0) != 0;
            cursor.close();
            return val;
        }
        cursor.close();
        return null;
    }

    public static String queryScalarString(SQLiteDatabase database, String sql)
    {
        Cursor cursor = database.rawQuery(sql, null);
        if (cursor.moveToFirst())
        {
            String val = cursor.isNull(0) ? null : cursor.getString(0);
            cursor.close();
            return val;
        }
        cursor.close();
        return null;
    }

    public static ContentValues getContentValuesForMap(Map<String, ?> values)
    {
        ContentValues content = new ContentValues();

        for (String key : values.keySet())
        {
            Object value = values.get(key);
            if (value == null)
            {
                content.putNull(key);
            }
            else if (value instanceof Integer)
            {
                content.put(key, (Integer)value);
            }
            else if (value instanceof Short)
            {
                content.put(key, (Short)value);
            }
            else if (value instanceof Long)
            {
                content.put(key, (Long)value);
            }
            else if (value instanceof Float)
            {
                content.put(key, (Float)value);
            }
            else if (value instanceof Double)
            {
                content.put(key, (Double)value);
            }
            else if (value instanceof Boolean)
            {
                content.put(key, (Boolean)value);
            }
            else if (value instanceof String)
            {
                content.put(key, (String)value);
            }
            else if (value instanceof Date)
            {
                content.put(key, getDateForDbFromDate((Date) value, false));
            }
        }

        return content;
    }

    public static String escape(String value)
    {
        return value.replace("'", "''");
    }
}

File: Helpers/src/main/java/com/dg/helpers/DbHelper.kt
package com.dg.helpers

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

@Suppress("unused")
object DbHelper
{
    fun utcIsoDateFormatter(): SimpleDateFormat
    {
        return DateHelper.utcIsoDateFormatter()
    }

    @Suppress("MemberVisibilityCanBePrivate")
    fun getDateForDbFromDate(date: Date?, escape: Boolean): String?
    {
        return if (date != null)
        {
            if (escape)
            {
                "'" + DateHelper.utcIsoDateFormatter().format(date) + "'"
            }
            else
            {
                DateHelper.utcIsoDateFormatter().format(date)
            }
        }
        else
        {
            if (escape)
            {
                "null"
            }
            else
            {
                null
            }
        }
    }

    fun getDateFromDbString(date: String?): Date?
    {
        if (date == null)
        {
            return null
        }
        try
        {
            return DateHelper.utcIsoDateFormatter().parse(date)
        }
        catch (ignored: ParseException)
        {
            return null
        }

    }

    fun queryScalarFloat(database: SQLiteDatabase, sql: String): Float?
    {
        val cursor = database.rawQuery(sql, null)
        if (cursor.moveToFirst())
        {
            val `val` = if (cursor.isNull(0)) null else cursor.getFloat(0)
            cursor.close()
            return `val`
        }
        cursor.close()
        return null
    }

    fun queryScalarInt(database: SQLiteDatabase, sql: String): Int?
    {
        val cursor = database.rawQuery(sql, null)
        if (cursor.moveToFirst())
        {
            val `val` = if (cursor.isNull(0)) null else cursor.getInt(0)
            cursor.close()
            return `val`
        }
        cursor.close()
        return null
    }

    fun queryScalarLong(database: SQLiteDatabase, sql: String): Long?
    {
        val cursor = database.rawQuery(sql, null)
        if (cursor.moveToFirst())
        {
            val `val` = if (cursor.isNull(0)) null else cursor.getLong(0)
            cursor.close()
            return `val`
        }
        cursor.close()
        return null
    }

    fun queryScalarDouble(database: SQLiteDatabase, sql: String): Double?
    {
        val cursor = database.rawQuery(sql, null)
        if (cursor.moveToFirst())
        {
            val `val` = if (cursor.isNull(0)) null else cursor.getDouble(0)
            cursor.close()
            return `val`
        }
        cursor.close()
        return null
    }

    fun queryScalarShort(database: SQLiteDatabase, sql: String): Short?
    {
        val cursor = database.rawQuery(sql, null)
        if (cursor.moveToFirst())
        {
            val `val` = if (cursor.isNull(0)) null else cursor.getShort(0)
            cursor.close()
            return `val`
        }
        cursor.close()
        return null
    }

    fun queryScalarBoolean(database: SQLiteDatabase, sql: String): Boolean?
    {
        val cursor = database.rawQuery(sql, null)
        if (cursor.moveToFirst())
        {
            val `val` = if (cursor.isNull(0)) null else cursor.getInt(0) != 0
            cursor.close()
            return `val`
        }
        cursor.close()
        return null
    }

    fun queryScalarString(database: SQLiteDatabase, sql: String): String?
    {
        val cursor = database.rawQuery(sql, null)
        if (cursor.moveToFirst())
        {
            val `val` = if (cursor.isNull(0)) null else cursor.getString(0)
            cursor.close()
            return `val`
        }
        cursor.close()
        return null
    }

    fun getContentValuesForMap(values: Map<String, *>): ContentValues
    {
        val content = ContentValues()

        for (key in values.keys)
        {
            when (val value = values[key])
            {
                null -> content.putNull(key)
                is Int -> content.put(key, value)
                is Short -> content.put(key, value)
                is Long -> content.put(key, value)
                is Float -> content.put(key, value)
                is Double -> content.put(key, value)
                is Boolean -> content.put(key, value)
                is String -> content.put(key, value)
                is Date -> content.put(key, getDateForDbFromDate(value, false))
            }
        }

        return content
    }

    fun escape(value: String): String
    {
        return value.replace("'", "''")
    }
}
